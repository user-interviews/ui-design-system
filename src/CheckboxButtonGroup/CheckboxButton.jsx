import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import FormControlLabel from '../FormControlLabel';

import './CheckboxButton.scss';

const CheckboxButton = ({
  bordered,
  checked,
  className,
  disabled,
  label,
  id,
  name,
  value,
  onChange,
  ...rest
}) => (
  <FormControlLabel
    className={classnames('CheckboxButton', {
      'CheckboxButton--bordered': bordered,
      'CheckboxButton--active': bordered && checked,
    })}
    htmlFor={id}
    labelHtmlFor={id}
    text={label}
  >
    <input
      checked={checked}
      className={className}
      disabled={disabled}
      id={id}
      name={name}
      type="checkbox"
      value={value}
      onChange={onChange}
      {...rest}
    />
  </FormControlLabel>
);

CheckboxButton.propTypes = {
  bordered: PropTypes.bool,
  checked: PropTypes.bool,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  name: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]),
  onChange: PropTypes.func,
};

CheckboxButton.defaultProps = {
  bordered: false,
  checked: undefined,
  className: undefined,
  disabled: undefined,
  name: undefined,
  value: undefined,
  onChange: undefined,
};

export default CheckboxButton;
