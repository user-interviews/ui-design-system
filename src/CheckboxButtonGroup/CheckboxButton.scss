@import '../../scss/theme';

.CheckboxButton {
  border: 1px solid $ux-gray-400;
  border-radius: $ux-border-radius;
  cursor: pointer;
  padding: 0.3125rem 0.625rem;
  margin-right: 0.5rem;

  display: flex;
  align-items: center;

  &__input {
    margin-right: 0.5rem;
  }

  &__label {
    @include font-type-30
  }

  &--active {
    background-color: $ux-blue-100;
  }
}
