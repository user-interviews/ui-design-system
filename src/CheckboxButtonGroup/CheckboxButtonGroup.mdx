import { ArgsTable, Story, Canvas } from '@storybook/addon-docs/blocks';

# Checkbox Button Group

##

<h4>
  A FormGroup that utilizes Checkboxes. Supports borders, row/column orientations, and descriptions.
</h4>

<Canvas>
  <Story id="design-system-form-elements-checkboxbuttongroup--default"/>
</Canvas>

### When to use 
- Reason 1
- Reason 2

### When to not use
- Reason 1
- Reason 2

## Props
- WIP

## Stories

### Default

- Multi-selection button. Can be toggled on/off by clicking.

<Canvas>
  <Story id="design-system-form-elements-checkboxbuttongroup--default"/>
</Canvas>

### Bordered

- Multi-selection button. Only used in the context of Checkbox FormGroups.

<Canvas>
  <Story id="design-system-form-elements-checkboxbuttongroup--bordered-row"/>
</Canvas>

### Description

- Multi-selection button. Includes helper text and/or a description in addition to the default label.

<Canvas>
  <Story id="design-system-form-elements-checkboxbuttongroup--description-row"/>
</Canvas>

## Formatting

### States

### Anatomy 

### Sizing

### Alignment


## Best practices

### General

- Required checkboxes (e.g. Terms of Service agreement, phone verification) should be unselected by default
- When using Default checkbox buttons, the options should always be stacked vertically (except when used in the context of Inline FormGroups)

### Behavior

### Implementation

### UX Copy
