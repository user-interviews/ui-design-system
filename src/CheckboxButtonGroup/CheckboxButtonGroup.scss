@import '../../scss/theme';
@import '../../scss/forms/control_button_group';

.CheckboxButtonGroup {
  display: flex;
  flex-direction: column;

  label {
    margin-bottom: $synth-spacing-2;
  }

  &--row {
   @include row-control-group;
  }

  &--column--full-width {
    width: 100%;
  }
}
