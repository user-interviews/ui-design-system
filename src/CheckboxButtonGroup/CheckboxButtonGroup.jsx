import React, { Children } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { ORIENTATIONS } from '../RadioButtonGroup/RadioButtonGroup';

import './CheckboxButtonGroup.scss';

export default function CheckboxButtonGroup({
  children,
  fullWidth,
  id,
  orientation,
  parseInput,
  value,
  onChange,
}) {
  const row = orientation === ORIENTATIONS.ROW;

  const handleChangeValues = (event) => {
    const eventValue = parseInput(event.target.value);
    const values = [...value];

    if (event.target.checked && !values.includes(eventValue)) {
      values.push(eventValue);
    } else if (!event.target.checked && values.includes(eventValue)) {
      values.splice(values.indexOf(eventValue), 1);
    }

    if (onChange) {
      onChange(values);
    }
  };

  const renderChildElement = (child) => {
    const { value: childValue } = child.props;
    const checked = value && value.includes(childValue);

    // Treat children as controlled components only if group is also controlled
    const childProps = onChange ? {
      checked,
      onChange: handleChangeValues,
    } : {};

    if (row) {
      childProps.bordered = true;
    }

    return React.cloneElement(child, childProps);
  };

  return (
    <div
      className={classNames('CheckboxButtonGroup', {
        'CheckboxButtonGroup--row': row,
        'CheckboxButtonGroup--row--full-width': row && fullWidth,
        'CheckboxButtonGroup--row--compact': row && !fullWidth,
      })}
      id={id}
    >
      {
        Children.toArray(children).map(renderChildElement)
      }
    </div>
  );
}

CheckboxButtonGroup.propTypes = {
  children: PropTypes.arrayOf(PropTypes.element).isRequired,
  fullWidth: PropTypes.bool,
  id: PropTypes.string.isRequired,
  orientation: PropTypes.oneOf(Object.values(ORIENTATIONS)),
  parseInput: PropTypes.func,
  value: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.string,
    ]),
  ),
  onChange: PropTypes.func,
};

CheckboxButtonGroup.defaultProps = {
  fullWidth: false,
  orientation: ORIENTATIONS.ROW,
  parseInput: (i) => i,
  value: [],
  onChange: undefined,
};
