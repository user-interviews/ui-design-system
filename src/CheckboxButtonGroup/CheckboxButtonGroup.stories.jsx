import React, { useState } from 'react';
import {
 boolean, radios, text, withKnobs,
} from '@storybook/addon-knobs';
import { ORIENTATIONS } from '../ControlButtonGroup';
import CheckboxButtonGroup from './index';
import CheckboxButton from '../CheckboxButton';
import FormControlLabel from '../FormControlLabel';
import FormGroup from '../FormGroup/FormGroup';
import mdx from './CheckboxButtonGroup.mdx';

export default {
  title: 'Components/Form Elements/CheckboxButtonGroup',
  component: CheckboxButtonGroup,
  decorators: [withKnobs()],
  parameters: {
    docs: {
      page: mdx,
    },
  },
};

/* eslint-disable react/prop-types */
const CheckboxButtonGroupComponent = ({
  bordered,
  children,
  defaultValue,
  fullWidth,
  id,
  inline,
  inputType,
  label,
  labelHelperText,
  orientation,
  labelHtmlFor,
}) => {
  const [value, setValue] = useState(defaultValue);

  return (
    <FormGroup
      bordered={bordered}
      id={id}
      inline={inline}
      inputType={inputType}
      label={label}
      labelHelperText={labelHelperText}
      labelHtmlFor={labelHtmlFor}
    >
      <CheckboxButtonGroup
        fullWidth={fullWidth}
        id={`${id}-button-group`}
        orientation={orientation}
        value={value}
        onChange={setValue}
      >
        { children }
      </CheckboxButtonGroup>
    </FormGroup>
  );
};
/* eslint-enable react/prop-types */

export const Default = () => {
  const orientation = radios(
    'Orientation',
    Object.values(ORIENTATIONS),
    ORIENTATIONS.COLUMN,
  );
  const bordered = boolean('Bordered buttons', false);

  return (
    <CheckboxButtonGroupComponent
      bordered={boolean('Bordered Form Group', false)}
      defaultValue={[]}
      fullWidth={boolean('Full width', false)}
      id="with-checkbox-button-group"
      inputType="checkbox"
      label="Label"
      labelHelperText="use the knobs to try out different variations"
      labelHtmlFor="checkbox-button-group"
      orientation={orientation}
    >
      <FormControlLabel
        bordered={bordered}
        Control={CheckboxButton}
        id="value-1"
        text={text('Label 1', 'Value 1')}
        value="1"
      />
      <FormControlLabel
        bordered={bordered}
        Control={CheckboxButton}
        id="value-2"
        text={text('Label 2', 'Value 2')}
        value="2"
      />
      <FormControlLabel
        bordered={bordered}
        Control={CheckboxButton}
        id="value-3"
        text={text('Label 3', 'Value 3')}
        value="3"
      />
    </CheckboxButtonGroupComponent>
  );
};

export const DefaultRow = () => (
  <CheckboxButtonGroupComponent
    bordered={false}
    defaultValue={[]}
    fullWidth={false}
    id="with-checkbox-button-group-1"
    inline
    inputType="checkbox"
    label="Label"
    labelHelperText="helper text"
    labelHtmlFor="checkbox-button-group"
    orientation={ORIENTATIONS.ROW}
  >
    <FormControlLabel
      bordered={false}
      Control={CheckboxButton}
      id="value-1-1"
      text="Label 1"
      value="1"
    />
    <FormControlLabel
      bordered={false}
      Control={CheckboxButton}
      id="value-2-1"
      text="Label 2"
      value="2"
    />
    <FormControlLabel
      bordered={false}
      Control={CheckboxButton}
      id="value-3-1"
      text="Label 3"
      value="3"
    />
  </CheckboxButtonGroupComponent>
  );

export const BorderedRow = () => (
  <CheckboxButtonGroupComponent
    bordered={false}
    defaultValue={[]}
    fullWidth={false}
    id="with-checkbox-button-group-2"
    inputType="checkbox"
    label="Label"
    labelHelperText="helper text"
    labelHtmlFor="checkbox-button-group"
    orientation={ORIENTATIONS.ROW}
  >
    <FormControlLabel
      bordered
      Control={CheckboxButton}
      id="value-1-2"
      text="Label 1"
      value="1"
    />
    <FormControlLabel
      bordered
      Control={CheckboxButton}
      id="value-2-2"
      text="Label 2"
      value="2"
    />
    <FormControlLabel
      bordered
      Control={CheckboxButton}
      id="value-3-2"
      text="Label 3"
      value="3"
    />
  </CheckboxButtonGroupComponent>
);

export const Description = () => (
  <CheckboxButtonGroupComponent
    bordered={false}
    defaultValue={[]}
    fullWidth
    id="with-checkbox-button-group-3"
    inputType="checkbox"
    label="Label"
    labelHelperText="helper text"
    labelHtmlFor="checkbox-button-group"
    orientation={ORIENTATIONS.COLUMN}
  >
    <FormControlLabel
      bordered
      Control={CheckboxButton}
      helperText="helper text"
      id="value-1-3"
      text="Label 1"
      value="1"
    >
      This is where the description goes
    </FormControlLabel>
    <FormControlLabel
      bordered
      Control={CheckboxButton}
      helperText="helper text"
      id="value-2-3"
      text="Label 2"
      value="2"
    >
      This is where the description goes
    </FormControlLabel>
    <FormControlLabel
      bordered
      Control={CheckboxButton}
      helperText="helper text"
      id="value-3-3"
      text="Label 3"
      value="3"
    >
      This is where the description goes
    </FormControlLabel>
  </CheckboxButtonGroupComponent>
);

export const DescriptionRow = () => (
  <CheckboxButtonGroupComponent
    bordered={false}
    defaultValue={[]}
    fullWidth
    id="with-checkbox-button-group-4"
    inputType="checkbox"
    label="Label"
    labelHelperText="helper text"
    labelHtmlFor="checkbox-button-group"
    orientation={ORIENTATIONS.ROW}
  >
    <FormControlLabel
      bordered
      Control={CheckboxButton}
      helperText="helper text"
      id="value-1-4"
      text="Label 1"
      value="1"
    >
      This is where the description goes
    </FormControlLabel>
    <FormControlLabel
      bordered
      Control={CheckboxButton}
      helperText="helper text"
      id="value-2-4"
      text="Label 2"
      value="2"
    >
      This is where the description goes
    </FormControlLabel>
    <FormControlLabel
      bordered
      Control={CheckboxButton}
      helperText="helper text"
      id="value-3-4"
      text="Label 3"
      value="3"
    >
      This is where the description goes
    </FormControlLabel>
  </CheckboxButtonGroupComponent>
);
