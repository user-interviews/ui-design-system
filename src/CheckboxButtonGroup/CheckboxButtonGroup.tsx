import React from 'react';
import classNames from 'classnames';

import ControlButtonGroup, { ORIENTATIONS } from '../ControlButtonGroup';

import './CheckboxButtonGroup.scss';

type CheckboxButtonGroupProps = {
  children: React.ReactElement[];
  fullWidth?: boolean;
  id: string;
  orientation?: unknown[];
  parseInput?: (arg0: string) => string | number;
  value?: (number | string)[];
  onChange?: (...args: unknown[]) => unknown;
};

export default function CheckboxButtonGroup({
  children,
  fullWidth = false,
  id,
  orientation = ORIENTATIONS.ROW,
  parseInput = (val) => val,
  value = [],
  onChange = undefined,
}: CheckboxButtonGroupProps) {
  const row = orientation === ORIENTATIONS.ROW;
  const column = orientation === ORIENTATIONS.COLUMN;

  const handleChangeValue = (event: React.ChangeEvent<HTMLInputElement>) => {
    const eventValue = parseInput(event.target.value);
    const values = Array.isArray(value) ? [...value] : [value];

    if (event.target.checked && !values.includes(eventValue)) {
      values.push(eventValue);
    } else if (!event.target.checked && values.includes(eventValue)) {
      values.splice(values.indexOf(eventValue), 1);
    }

    if (onChange) {
      onChange(values);
    }
  };

  const childChecked = (childValue) => (
    Array.isArray(value) && value.includes(childValue)
  );

  return (
    <div
      className={classNames('CheckboxButtonGroup', {
        'CheckboxButtonGroup--row': row,
        'CheckboxButtonGroup--row--full-width': row && fullWidth,
        'CheckboxButtonGroup--row--compact': row && !fullWidth,
        'CheckboxButtonGroup--column--full-width': column && fullWidth,
      })}
      id={id}
    >
      <ControlButtonGroup
        childChecked={childChecked}
        handleChangeValue={handleChangeValue}
        orientation={orientation}
        onChange={onChange}
      >
        {children}
      </ControlButtonGroup>
    </div>
  );
}
