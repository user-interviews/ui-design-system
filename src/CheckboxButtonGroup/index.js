import CheckboxButtonGroup from './CheckboxButtonGroup';

export default CheckboxButtonGroup;
