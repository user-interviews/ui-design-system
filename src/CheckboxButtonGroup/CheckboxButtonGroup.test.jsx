import React, { useState } from 'react';
import { act, create } from 'react-test-renderer';
import PropTypes from 'prop-types';

import CheckboxButtonGroup from 'src/CheckboxButtonGroup';
import CheckboxButton from 'src/CheckboxButton';

const CheckboxButtonGroupComponent = ({ children, defaultValues }) => {
  const [value, setValue] = useState(defaultValues);
  return (
    <CheckboxButtonGroup id="checkbox-question" value={value} onChange={setValue}>
      {children}
    </CheckboxButtonGroup>
  );
};

CheckboxButtonGroupComponent.propTypes = {
  defaultValues: PropTypes.array.isRequired,
};

describe('CheckboxButtonGroup', () => {
  test('sets initial value properly', () => {
    const checkboxButtonGroup = create(
      <CheckboxButtonGroupComponent defaultValues={[1, 2]}>
        <CheckboxButton id="first" label="First value" value={1} />
        <CheckboxButton id="second" label="Second value" value={2} />
      </CheckboxButtonGroupComponent>,
    );

    expect(checkboxButtonGroup).toMatchSnapshot();
  });

  test('will update array state to check a new item', async () => {
    const checkboxButtonGroup = create(
      <CheckboxButtonGroupComponent defaultValues={[]}>
        <CheckboxButton id="first" label="First value" value={1} />
        <CheckboxButton id="second" label="Second value" value={2} />
      </CheckboxButtonGroupComponent>,
    );

    const event = { target: { value: 1, checked: true } };
    const firstButton = checkboxButtonGroup.root.findAllByType((CheckboxButton))[0];

    await act(async () => firstButton.props.onChange(event));
    expect(checkboxButtonGroup).toMatchSnapshot();
  });

  test('will update array state to uncheck an item', async () => {
    const checkboxButtonGroup = create(
      <CheckboxButtonGroupComponent defaultValues={[1]}>
        <CheckboxButton id="first" label="First value" value={1} />
        <CheckboxButton id="second" label="Second value" value={2} />
      </CheckboxButtonGroupComponent>,
    );

    const event = { target: { value: 1, checked: false } };
    const firstButton = checkboxButtonGroup.root.findAllByType(CheckboxButton)[0];

    await act(async () => firstButton.props.onChange(event));
    expect(checkboxButtonGroup).toMatchSnapshot();
  });
});
