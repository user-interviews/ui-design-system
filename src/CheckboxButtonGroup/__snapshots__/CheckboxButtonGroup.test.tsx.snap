// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxButtonGroup sets initial value properly 1`] = `
<div
  className="CheckboxButtonGroup CheckboxButtonGroup--row CheckboxButtonGroup--row--compact"
  id="checkbox-question"
>
  <input
    bordered={true}
    checked={true}
    className=""
    id="first"
    label="First value"
    onChange={[Function]}
    type="checkbox"
    value={1}
  />
  <input
    bordered={true}
    checked={true}
    className=""
    id="second"
    label="Second value"
    onChange={[Function]}
    type="checkbox"
    value={2}
  />
</div>
`;

exports[`CheckboxButtonGroup will update array state to check a new item 1`] = `
<div
  className="CheckboxButtonGroup CheckboxButtonGroup--row CheckboxButtonGroup--row--compact"
  id="checkbox-question"
>
  <input
    bordered={true}
    checked={true}
    className=""
    id="first"
    label="First value"
    onChange={[Function]}
    type="checkbox"
    value={1}
  />
  <input
    bordered={true}
    checked={false}
    className=""
    id="second"
    label="Second value"
    onChange={[Function]}
    type="checkbox"
    value={2}
  />
</div>
`;

exports[`CheckboxButtonGroup will update array state to uncheck an item 1`] = `
<div
  className="CheckboxButtonGroup CheckboxButtonGroup--row CheckboxButtonGroup--row--compact"
  id="checkbox-question"
>
  <input
    bordered={true}
    checked={false}
    className=""
    id="first"
    label="First value"
    onChange={[Function]}
    type="checkbox"
    value={1}
  />
  <input
    bordered={true}
    checked={false}
    className=""
    id="second"
    label="Second value"
    onChange={[Function]}
    type="checkbox"
    value={2}
  />
</div>
`;
